### PR TITLE
Add Tenant Id Challenges Support.

### LINKED ISSUE #22353: Add support for tenant Id Challenges / tenant discovery in ClientCredentials
Currently, for services that return challenges which include context about the tenant that can service a failed service request, such as `authorization_url`,
 Azure.Identity does not utilize this information to request a new token with the correct tenantId.

 This is a proposal to enhance Azure.Identity to make it possible for service clients to interpret challenges to request tokens for the correct tenant Id, where possible.

The proposal details can be found in [this gist](https://gist.github.com/christothes/34b713a70aad4e94f228b5a9b64f92a4)

Related: https://github.com/Azure/azure-sdk-for-net/issues/20302

### PR DESCRIPTION
**Pull request by g2vinay**
Fixes #22353

### CODE REVIEW

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/TokenRequestContextExperimental.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Contains details of a request to get a token.
+ */
+public class TokenRequestContextExperimental extends TokenRequestContext {
```
Comment (srnagar) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/TokenRequestContextExperimental.java:16]: It would be good to give this a name that will eventually be used when it's moved to `azure-core`.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package com.azure.identity.implementation.util;
+
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.experimental.credential.TokenRequestContextExperimental;
+import com.azure.core.util.CoreUtils;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.identity.implementation.IdentityClientOptions;
+
+public class IdentityUtil {
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java:9]: Make this `final`. Add javadoc and include a private ctor as it doesn't require instantiation.

File: sdk/resourcemanager/demo/HELP.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Getting Started
```
Comment (srnagar) [inline sdk/resourcemanager/demo/HELP.md:1]: Why is this added?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    private static final ClientLogger logger = new ClientLogger(IdentityUtil.class);
+
+    /**
+     * Convert a resource to a list of scopes.
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java:13]: Fix javadoc

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        if (requestContext instanceof TokenRequestContextExperimental) {
+            contextTenantId = ((TokenRequestContextExperimental) requestContext).getTenantId();
+        } else {
+            return currentTenantId;
+        }
+
+        if (!options.getAllowMultiTenantAuthentication()) {
+            if (contextTenantId != null && currentTenantId != contextTenantId
+                && !options.isLegacyTenantSelectionEnabled()) {
+                logger.logExceptionAsError(new RuntimeException("The TenantId received from a challenge did not match "
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java:32]: Throw `IllegalStateException` instead of using `RuntimeException` directly.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 6]
```diff
@@ -53,6 +56,8 @@ public IdentityClientOptions() {
         retryTimeout = i -> Duration.ofSeconds((long) Math.pow(2, i.getSeconds() - 1));
         regionalAuthority = RegionalAuthority.fromString(
             configuration.get(Configuration.PROPERTY_AZURE_REGIONAL_AUTHORITY_NAME));
+        identityLegacyTenantSelection = configuration
+            .get(AZURE_IDENTITY_ENABLE_LEGACY_TENANT_SELECTION, false) ;
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:60]: Is this env var set by the user? If so, should we make this constant publicly available? It is public in this class but it's in implementation package.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 7]
```diff
@@ -197,10 +202,30 @@ public IdentityClientOptions setAllowUnencryptedCache(boolean allowUnencryptedCa
...
     public boolean getAllowUnencryptedCache() {
         return this.allowUnencryptedCache;
     }
 
+    /**
+     * Get the flag indicating if multi tenant authentication is enabled or not.
+     *
+     * @return the boolean status indicating if multi tenant authentication is enabled or not.
+     */
+    public boolean getAllowMultiTenantAuthentication() { return this.allowMultiTenantAuthentication; }
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:227]: fix formatting to not have the method body in the same line.

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/TokenRequestContextExperimental.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        super();
+    }
+
+    /**
+     * Set the tenant to be used in the authentication request.
+     *
+     * @param tenantId the tenant to be used when retrieving the token.
+     * @return the updated TokenRequestContext itself
+     */
+    public TokenRequestContext setTenantId(String tenantId) {
```
Comment (srnagar) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/TokenRequestContextExperimental.java:32]: This should return `TokenRequestContextExperimental` and also override all setters in the super class.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java (hunk 7 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:227]: Do you remember why is this and the method above not following `isXxx` naming convention for boolean properties?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java (hunk 2 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java:9]: Add a private constructor to prevent instantiation

**Review by jianghaolu — APPROVED** [review]
couple nits

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/TokenRequestContextExperimental.java (hunk 1 above)
Comment (g2vinay) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/TokenRequestContextExperimental.java:16]: This will merge with TokenRequestContext when GA happens

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+package com.azure.identity.implementation.util;
+
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.experimental.credential.TokenRequestContextExperimental;
+import com.azure.core.util.CoreUtils;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.identity.implementation.IdentityClientOptions;
+
+public final class IdentityUtil {
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityUtil.java:12]: nit: add javadoc for class

**Review by srnagar — APPROVED** [review]
